### PR TITLE
Feat: Add authentication routes

### PR DESCRIPTION
**Pull request by vic3king**
**What does this Merge request do?**

Adds Authentication routes

**Description of Task to be completed?**

- Folder structure
- DB setup
- Add registration endpoint
- Add login endpoint
- Setup Joi for data validation
- Setup email and add email verification on registration
- Configure babel
- Add bullJs for queuing emails
- Add password reset functionality
- Add endpoints to get and update a users profile

**How should this be manually tested?**

xxx

**Any background context you want to provide?**

- I added an email verification step to this... Thoughts?

**What are the relevant Trello cards/stories?**

xxx

**Screenshots (if appropriate)**

xxx

**Questions:**

xxx

### CODE REVIEW

**Comment by 9trocode** [comment]
@sojida @otseobande do review this as well

**Comment by 9trocode** [comment]
@vic3king about the email verification since we can't afford to pay for SendGrid think we should allow them him directly but I think SendGrid offers first 100 emails free if they still do we can stick with that else let's diable that features for now.

Or what do you think ??

**Comment by 9trocode** [comment]
@otseobande we need ci how soon can we get it ??

File: server/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+      const verifiedPassword = await authHelper.comparePassword(
+        password,
+        hashedPassword
+      );
+
+      if (!verifiedPassword) {
+        return res.status(401).send({
+          status: false,
+          error: 'Bad Login',
```
Comment (9trocode) [inline server/controllers/authController.js:115]: let's use the normal `invalid email or password convention` here

File: server/controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+      return res.status(200).json({
+        status: true,
+        message: 'Login successful',
+        token,
+        user,
+      });
+    } catch (error) {
+      return res.status(500).send({
+        status: false,
+        message: 'something went wrong',
```
Comment (9trocode) [inline server/controllers/authController.js:134]: Can you please add some synthetic sugar to this ??

File: server/database/seeders/20200113164607-User.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module.exports = {
+  up: (queryInterface) =>
+    queryInterface.bulkInsert('Users', [
+      {
+        id: '6517a6ea-662b-4eef-ab9f-20f89bd7099c',
+        fullName: 'John Doe',
+        username: 'jd',
+        email: 'john@gmail.com',
```
Comment (sojida) [inline server/database/seeders/20200113164607-User.js:8]: `john@gmail.com` could be someone's actual email, can we use `{nameOfUser}@example.com` instead?

File: server/controllers/authController.js (hunk 2 above)
Comment (vic3king) [reply server/controllers/authController.js:134]: @9trocode Please can explain what you mean? I don't understand the feedback.

File: server/controllers/authController.js (hunk 1 above)
Comment (vic3king) [reply server/controllers/authController.js:115]: Noted

**Comment by vic3king** [comment]
> @vic3king is it possible we write integration test for this feature ??
@9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?

**Comment by vic3king** [comment]
> @vic3king about the email verification since we can't afford to pay for SendGrid think we should allow them him directly but I think SendGrid offers first 100 emails free if they still do we can stick with that else let's diable that features for now.
> 
> Or what do you think ??

They still do @9trocode. I added that because typically chatting platforms do some form of user verifications for emails or phone numbers before they allow use

File: server/database/seeders/20200113164607-User.js (hunk 3 above)
Comment (vic3king) [reply server/database/seeders/20200113164607-User.js]: Sure. I'll update

File: server/controllers/authController.js (hunk 2 above)
Comment (9trocode) [reply server/controllers/authController.js:134]: in short, words I mean can we use something better for this e.g `Something went wrong while processing your request do try again later`

**Comment by 9trocode** [comment]
> > @vic3king about the email verification since we can't afford to pay for SendGrid think we should allow them him directly but I think SendGrid offers first 100 emails free if they still do we can stick with that else let's diable that features for now.
> > Or what do you think ??
> 
> They still do @9trocode. I added that because typically chatting platforms do some form of user verifications for emails or phone numbers before they allow use

Okay nice @vic3king

**Comment by 9trocode** [comment]
> > @vic3king is it possible we write integration test for this feature ??
> > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?

No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend

**Comment by vic3king** [comment]
> > > @vic3king is it possible we write integration test for this feature ??
> > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> 
> No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend

@9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?

**Comment by 9trocode** [comment]
> > > > @vic3king is it possible we write integration test for this feature ??
> > > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> > 
> > 
> > No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend
> 
> @9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?

Yea you can put it the way but a QA job is way more than that but anyways we need to also write our own backend test and frontend test if it won't be stressful though else we can just leave it to me I will write my postman test as well as integration test for frontend

**Comment by vic3king** [comment]
> > > > > @vic3king is it possible we write integration test for this feature ??
> > > > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> > > 
> > > 
> > > No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend
> > 
> > 
> > @9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?
> 
> Yea you can put it the way but a QA job is way more than that but anyways we need to also write our own backend test and frontend test if it won't be stressful though else we can just leave it to me I will write my postman test as well as integration test for frontend

I'll write the tests. Later today no worries.

**Comment by 9trocode** [comment]
> > > > > > @vic3king is it possible we write integration test for this feature ??
> > > > > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> > > > 
> > > > 
> > > > No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend
> > > 
> > > 
> > > @9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?
> > 
> > 
> > Yea you can put it the way but a QA job is way more than that but anyways we need to also write our own backend test and frontend test if it won't be stressful though else we can just leave it to me I will write my postman test as well as integration test for frontend
> 
> I'll write the tests. Later today no worries.

okay, should I leave this pr open or I merge it since it's ready ??

**Comment by vic3king** [comment]
> > > > > > > @vic3king is it possible we write integration test for this feature ??
> > > > > > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> > > > > 
> > > > > 
> > > > > No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend
> > > > 
> > > > 
> > > > @9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?
> > > 
> > > 
> > > Yea you can put it the way but a QA job is way more than that but anyways we need to also write our own backend test and frontend test if it won't be stressful though else we can just leave it to me I will write my postman test as well as integration test for frontend
> > 
> > 
> > I'll write the tests. Later today no worries.
> 
> okay, should I leave this pr open or I merge it since it's ready ??

merge and I'll raise another for tests.

**Comment by 9trocode** [comment]
> > > > > > > > @vic3king is it possible we write integration test for this feature ??
> > > > > > > > @9trocode I tried to clarify this over the call the other day. Is that not part of the QA job?
> > > > > > 
> > > > > > 
> > > > > > No, apart from the integration test writing inside the code a QA job is way above that part though for this test all I will be writing is integration test on UI using something like cypress to automate and I will also write smoke and regression test as well using postman test for backend
> > > > > 
> > > > > 
> > > > > @9trocode But these are mostly frontend tests. In my opinion, QA should cover these as well. Thoughts @otseobande ?
> > > > 
> > > > 
> > > > Yea you can put it the way but a QA job is way more than that but anyways we need to also write our own backend test and frontend test if it won't be stressful though else we can just leave it to me I will write my postman test as well as integration test for frontend
> > > 
> > > 
> > > I'll write the tests. Later today no worries.
> > 
> > 
> > okay, should I leave this pr open or I merge it since it's ready ??
> 
> merge and I'll raise another for tests.

okay then

**Pull request merged**